include upgraded packages in pacman upgrade action

yithian opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
pacman module

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (devel d4449d082e) last updated 2017/01/26 11:34:53 (GMT -400)
  config file = /home/<username>/code/<playbooks repo>/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Include a list of packages that were upgraded when performing a pacman upgrade task (or would be, if running in check mode). The upgraded packages are returned as a new key in the interest of not breaking backwards compatibility. Also this way there's less risk of flooding the user's screen with a huge list of packages.

I envision this being useful for registering the return of an upgrade so that handlers could selectively restart services after being upgraded.
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

<!-- Paste verbatim command output below, e.g. before and after your change -->
Before:
```
ok: [<hostname>] => {
    "updates": {
        "changed": true, 
        "msg": "System upgraded"
    }
}
```
After:
```
ok: [<hostname>] => {
    "updates": {
        "changed": true, 
        "msg": "System upgraded", 
        "packages": [
            "gnupg", 
            "python2-parsedatetime", 
            "python2-requests"
        ]
    }
}
```

yithian committed:
include upgraded packages in pacman upgrade action

indrajitr commented:
shipit

jctanner commented:
bot_status

bcoca commented:
I would use the 'diff' facility to show packages added/removed.

indrajitr commented:
Yes, that's on my todo-list. Will get back to this in a week.

indrajitr commented:
needs_revision
If @yithian can take care of this per the recommendation of @bcoca (which is definitely a better way forward), great. Else I'll take care of this in a week.

yithian commented:
I'll make that revision and update the pull request with the new commit

yithian committed:
display upgraded packages as diff output for pacman upgrade

yithian commented:
--check --diff output now looks like this:

```
TASK [system updates] ***********************************************************************************************************
--- before
+++ after: upgraded
@@ -0,0 +1,3 @@
+gnupg
+python2-parsedatetime
+python2-requests

changed: [<hostname>]

TASK [display output] ***********************************************************************************************************
ok: [<hostname>] => {
    "pacman_result": {
        "changed": true,
        "diff": {
            "after": "gnupg\npython2-parsedatetime\npython2-requests\n",
            "after_header": "upgraded",
            "before": ""
        },
        "msg": "3 package(s) would be upgraded",
        "packages": [
            "gnupg",
            "python2-parsedatetime",
            "python2-requests"
        ]
    }
}
```

using just --diff (without --check) now looks like this (probably redundant, but for completion's sake):

```
TASK [system updates] ***********************************************************************************************************
--- before
+++ after: upgraded
@@ -0,0 +1,3 @@
+gnupg
+python2-parsedatetime
+python2-requests

changed: [<hostname>]

TASK [display output] ***********************************************************************************************************
ok: [<hostname>] => {
    "pacman_result": {
        "changed": true,
        "diff": {
            "after": "gnupg\npython2-parsedatetime\npython2-requests\n",
            "after_header": "upgraded",
            "before": ""
        },
        "msg": "System upgraded",
        "packages": [
            "gnupg",
            "python2-parsedatetime",
            "python2-requests"
        ]
    }
}
```

Running without --diff will have output like from before but with a skeleton diff key. I left the packages key because I figured it would be easier to parse

yithian commented:
That was an accidental click of the close button, sorry for the noise

jtyr commented on lib/ansible/modules/packaging/os/pacman.py:
```diff
@@ -191,18 +191,26 @@ def upgrade(module, pacman_path):
             module.fail_json(msg="Could not upgrade")
     else:
-        module.exit_json(changed=False, msg='Nothing to upgrade')
+        module.exit_json(changed=False, msg='Nothing to upgrade', packages=data)
```
Any non-standard exit keywords should be [documented](http://docs.ansible.com/ansible/dev_guide/developing_modules.html#example) in the `RETURN` block.

yithian committed:
document the packages return key

yithian commented:
I've also got a patch that would add --diff support to the install_packages, remove_packages and check_packages functions, if that should be included here.

If not, I can submit it as a separate pull request.

indrajitr commented:
@yithian, yes please! That would be very awesome. I'll just take your PR instead of submitting the one I was working on.

yithian committed:
add --diff support for installing, removing and checking packages

yithian commented:
output used to look like:
```
TASK [install packages] ********************************************************
changed: [<hostname>] => (item=[u'cowsay', u'ponysay'])

TASK [remove packages] *********************************************************
changed: [<hostname>] => (item=[u'cowsay', u'ponysay'])
```

output now looks like:
```
TASK [install packages] *********************************************************************************************************************************************************************************************
--- before
+++ after: installed
@@ -0,0 +1,2 @@
+cowsay
+ponysay

changed: [<hostname>] => (item=[u'cowsay', u'ponysay'])

TASK [remove packages] **********************************************************************************************************************************************************************************************
--- before: removed
+++ after
@@ -1,2 +0,0 @@
-cowsay
-ponysay

changed: [<hostname>] => (item=[u'cowsay', u'ponysay'])
```

yithian commented:
any other changes i should make for this?

yithian commented:
ready_for_review

yithian commented:
bot_status

yithian commented:
bot_broken

yithian committed:
include package versions in pacman diff output

yithian commented:
upgrade output (with -D) now looks like:
```
--- before
+++ after
@@ -1,8 +1,8 @@
-bash-4.4.005-1
-libssh-0.7.4-1
-mesa-17.0.0-1.1
-libgl-17.0.0-1.1
-libwww-6.15-1
-appdirs-1.4.0-5
-utils-1.10.1-1
-wget-1.19.1-1
+bash-4.4.012-2
+libssh-0.7.4-2
+mesa-17.0.0-2
+libgl-17.0.0-2
+libwww-6.19-2
+appdirs-1.4.1-1
+utils-1.12.3-1
+wget-1.19.1-2

changed: [<hostname>] => {
    "changed": true, 
    "diff": {
        "after": "bash-4.4.012-2\nlibssh-0.7.4-2\nmesa-17.0.0-2\nlibgl-17.0.0-2\nlibwww-6.19-2\nappdirs-1.4.1-1\nutils-1.12.3-1\nwget-1.19.1-2\n", 
        "before": "bash-4.4.005-1\nlibssh-0.7.4-1\nmesa-17.0.0-1.1\nlibgl-17.0.0-1.1\nlibwww-6.15-1\nappdirs-1.4.0-5\nutils-1.10.1-1\nwget-1.19.1-1\n"
    }, 
    "invocation": {
        "module_args": {
            "force": false, 
            "name": null, 
            "recurse": false, 
            "state": "present", 
            "update_cache": true, 
            "upgrade": true
        }, 
        "module_name": "pacman"
    }, 
    "msg": "System upgraded", 
    "packages": [
        "bash", 
        "libssh", 
        "mesa", 
        "libgl", 
        "libwww", 
        "appdirs", 
        "utils", 
        "wget"
    ]
}
```

package install output (with -D) now looks like:
```
--- before
+++ after
@@ -0,0 +1,3 @@
+cowsay-3.03-9
+ponysay-3.0.2-1
+sl-5.02-4

changed: [<hostname>] => (item=[u'cowsay', u'ponysay', u'sl']) => {
    "changed": true, 
    "diff": {
        "after": "cowsay-3.03-9\nponysay-3.0.2-1\nsl-5.02-4\n", 
        "before": ""
    }, 
    "invocation": {
        "module_args": {
            "force": false, 
            "name": [
                "cowsay", 
                "ponysay", 
                "sl"
            ], 
            "recurse": false, 
            "state": "present", 
            "update_cache": false, 
            "upgrade": false
        }, 
        "module_name": "pacman"
    }, 
    "item": [
        "cowsay", 
        "ponysay", 
        "sl"
    ], 
    "msg": "installed 3 package(s). "
}
```

package removal output (with -D) now looks like:
```
--- before
+++ after
@@ -1,3 +0,0 @@
-cowsay-3.03-9
-ponysay-3.0.2-1
-sl-5.02-4

changed: [<hostname>] => (item=[u'cowsay', u'ponysay', u'sl']) => {
    "changed": true, 
    "diff": {
        "after": "", 
        "before": "cowsay-3.03-9\nponysay-3.0.2-1\nsl-5.02-4\n"
    }, 
    "invocation": {
        "module_args": {
            "force": false, 
            "name": [
                "cowsay", 
                "ponysay", 
                "sl"
            ], 
            "recurse": false, 
            "state": "absent", 
            "update_cache": false, 
            "upgrade": false
        }, 
        "module_name": "pacman"
    }, 
    "item": [
        "cowsay", 
        "ponysay", 
        "sl"
    ], 
    "msg": "removed 3 package(s)"
}
```

yithian commented:
ready_for_review

yithian commented:
@indrajitr @bcoca Is there anything left you'd like me to do for this?

It doesn't look like the ansibot commands are working to remove the needs_revision label...

jctanner commented:
@yithian the needs_revision label was set because @jtyr needed to approve or dismiss his change request. The bot status command wasn't working because of https://github.com/ansible/ansibullbot/pull/408
